src/skmultiflow/anomaly_detection/half_space_trees.py

```diff
@@ -0,0 +1,557 @@
+
+        classes: numpy.ndarray, optional (default=None)
+            Array with all possible/known class labels. This is an optional parameter, except
+            for the first partial_fit call where it is compulsory.
+
+        sample_weight: numpy.ndarray of shape (n_samples), optional (default=None)
+            Samples weight. If not provided, uniform weights are assumed.
+
+        Returns
+        -------
+        self
+
+        """
+
+        row_cnt, _ = X.shape
+
+        if self.samples_seen == 0:
+            self.build_trees()
+
+        for i in range(row_cnt):
+            self._partial_fit(X[i], y[i])
+
+        return self
+
+    def _partial_fit(self, X, y):
+        """ Trains the model on samples X and corresponding targets y.
+
+        Private function where actual training is carried on.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+
+        y: int
+            Class label for sample X.
+
+        """
+
+        # Populates first reference mass profile for every tree defined in the ensemble.
+        if self.samples_seen < self.window_size:
+            self.update_mass(X, True)
+        # Otherwise populates latest mass profile for every tree defined in the ensemble.
+        else:
+            self.update_mass(X, False)
+
+        if self.is_learning_phase_on is True and self.samples_seen > self.window_size:
+            self.set_is_learning_phase_on(False)
+
+        if (self.samples_seen % self.window_size) == 0:
+            self.update_models()
+
+        self.samples_seen += 1
+
+    def predict(self, X):
+        """ Predict classes for the passed data.
+
+        Parameters
+        ----------
+        X : numpy.ndarray of shape (n_samples, n_features)
+            The set of data samples to predict the class labels for.
+
+        Returns
+        -------
+        A numpy.ndarray with all the predictions for the samples in X.
+
+        """
+        r, _ = get_dimensions(X)
+        predictions = []
+        for i in range(r):
+            votes = self.get_votes_for_instance(X[i])
+            if votes == {}:
+                # Ensemble is empty, all classes equal, default to zero
+                predictions.append(0)
+            else:
+                predictions.append(max(votes, key=votes.get))
+        return np.asarray(predictions)
+
+    def get_votes_for_instance(self, X):
+        """ Get class votes for a single instance.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of length equal to the number of features.
+            Instance attributes.
+
+        Returns
+        -------
+        dict (class_value, weight)
+
+        """
+
+        combined_votes = {}
+        max_score = self.window_size * pow(2.0, self.depth)
+
+        for i in range(self.n_estimators):
+            vote = self.ensemble[i].get_votes_for_instance(X, max_score)
```

Comment:
This line partially duplicates the functionality of `predict_proba()` and leaves the internal `votes` values exposed. My suggestion is to change this line to:

```python
            vote = self.ensemble[i].predict_proba(X)
```

**Note 1:** This way you will get an array instead of a dictionary.
**Note 2:** With this change, `max_score` in no longer required.
**Note 3:** This way the `predict` and `predict_proba` are related (as desired) and potential bugs can be easily identified if either method is impacted by future (unrelated) changes.

Reply:
> This line partially duplicates the functionality of `predict_proba()` and leaves the internal `votes` values exposed. My suggestion is to change this line to:
> 
> ```python
>             vote = self.ensemble[i].predict_proba(X)
> ```
> 
> **Note 1:** This way you will get an array instead of a dictionary.
> **Note 2:** With this change, `max_score` in no longer required.
> **Note 3:** This way the `predict` and `predict_proba` are related (as desired) and potential bugs can be easily identified if either method is impacted by future (unrelated) changes.

I agree with all points. I guess you meant:
```
            vote = self.predict_proba(X)
```
Which makes use of the method `def predict_proba(self, X):` in class HalfSpaceTrees(which is not used).

Reply:
Using `self.predict_proba` is another option. 

Notice that in both cases, you could remove completely this method and move the functionality to `self.predict`. The thing is that `self.predict_proba` and `self.ensemble[i].predict_proba(X)` support multiple samples in `X` whereas this function process one sample at a time (since it is called inside a `for` loop).

I am in favor of using `self.predict_proba` since it will make the code more readable.